### PR TITLE
[DCA] Use `Leases` for `leaderelection` when they are available

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?
From K8s 1.26, the leader election package doesn’t support anymore using a configmap/endpoint object to store the leader token. This PR uses Leases to store the leaderelection token. Since Leases were added in k8s 1.14, we add support for ConfigMapsLock with a custom configmapslock implementation.

### Motivation
We need to continue upgrade to recent kubernetes libraries version to use a supported version for bug and security fixes but we also need to support old kubernetes versions. We need to implement this fix to bump k8s versions.

### Additional Notes
Merged in https://github.com/DataDog/datadog-agent/pull/16652 but reverted because of RBAC issues. Helm-chart and Operator have been updated.

Needs to be merged https://github.com/DataDog/helm-charts/pull/1108

### Possible Drawbacks / Trade-offs
Requires new RBACs to `list`, `watch`, `get` and `create` leases.

### Describe how to test/QA your changes
Testing requires a Kubernetes cluster with an old version (<= 1.14) and another one with a recent version.
You can deploy multiple cluster-agent and verify that they agree on the same leader after a few seconds. You can also delete the pod of the leader or add new replicas and verify that every cluster-agent views the same leader.
To find the leader for a given pod, you can execute kubectl exec cluster-agent-pod -- agent status | grep -i leader.

In order to deploy with an old kubernetes cluster, you will need to install a VM. If possible, with an architecture different from ARM. Then you will need to install an old version of minikube and docker if you do not use arm. I tested it with [minikube v1.13.0](https://github.com/kubernetes/minikube/releases/tag/v1.13.0). Finally, deploy your cluster with:
minikube start --kubernetes-version=v1.13.2 --driver=docker|none.
You can use the command kubectl api-resources to check if leases.coordination/v1 are available. They should not be.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: releasenotes/notes/use-leases-leader-election-c278c71fb39c2c84.yaml
Comment (cswatt) [inline releasenotes/notes/use-leases-leader-election-c278c71fb39c2c84.yaml:4]:
```suggestion
    Use ``leases.coordination/v1`` for Cluster Agent leader election when they are available.
```

**Review by cswatt — APPROVED** [review]
left comment on release note, approving so as not to block

File: releasenotes/notes/use-leases-leader-election-c278c71fb39c2c84.yaml
Comment (AliDatadog) [reply releasenotes/notes/use-leases-leader-election-c278c71fb39c2c84.yaml]: Done ! Thank you

File: pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go
Comment (clamoriniere) [inline pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go:150]: Since the addition of missing leases RBAC in the Helm chart are quite new, and the fact that the datadog-operator supporting it is not yet released.  I think it is safer to keep `configmap` the default object use for leader election. 

My idea is to add a new parameter in the config package (near https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config.go#L715-L717) to default the object to `configmap` 

```golang
config.BindEnvAndSetDefault("leader_election_default_resource","configmap")
```

We keep this option defaulted to configmap for a while. and then we can empty the default value. 
The behaviour that I think make sense is 

* `if leader_election_default_resource == "configmap"` use `configmap` without checking if "lease" object is present
* `if leader_election_default_resource == "lease"`  use "lease" without checking if "lease" object is present
* `if leader_election_default_resource == ""` check the present of "lease" resource, if present use it, else use configmap.


By doing this it will allow us to be backward compatible, but also allow to default to "" in the helm chart to use the new logique.

Comment (AliDatadog) [reply pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go]: It's now done. I'll open a PR to update our helm chart

**Comment by AliDatadog** [comment]
> left comment on release note, approving so as not to block

I edited the release notes 🙇

**Comment by AliDatadog** [comment]
@clamoriniere we also have a public helm chart update with https://github.com/DataDog/helm-charts/pull/1108

File: pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go
Comment (clamoriniere) [inline pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go:289]: maybe the variable name can be more self explained
for example 
```suggestion
resourceType := config.Datadog.GetString("leader_election_default_resource")
```

File: pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go
Comment (clamoriniere) [inline pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go:290]: I think it is better to compare the string instead of using `string.Contains()` because with `contains` string like `configmap,lease` will match

**Review by clamoriniere — APPROVED** [review]
Looks good to please.
please add an entry in the CODEOWNERS file for the new packages

File: .github/CODEOWNERS
Comment (pgimalac) [inline .github/CODEOWNERS:431]:
```suggestion
/internal/third_party/client-go         @DataDog/container-integrations
/internal/third_party/kubernetes        @DataDog/container-integrations
```

**Review by pgimalac — APPROVED** [review]
LGTM for team shared-components, just a typo in Codeowners file

File: releasenotes-dca/notes/use-leases-leader-election-c278c71fb39c2c84.yaml
Comment (buraizu) [inline releasenotes-dca/notes/use-leases-leader-election-c278c71fb39c2c84.yaml:5]:
```suggestion
    If this parameter is empty, leader election automatically detects if leases
```

File: releasenotes-dca/notes/use-leases-leader-election-c278c71fb39c2c84.yaml
Comment (buraizu) [inline releasenotes-dca/notes/use-leases-leader-election-c278c71fb39c2c84.yaml:4]:
```suggestion
    ``leader_election_default_resource`` to ``leases``, available since Kubernetes version 1.14. 
```

File: releasenotes-dca/notes/use-leases-leader-election-c278c71fb39c2c84.yaml
Comment (buraizu) [inline releasenotes-dca/notes/use-leases-leader-election-c278c71fb39c2c84.yaml:8]:
```suggestion
    Kubernetes versions previous to 1.14.
```

**Review by buraizu — APPROVED** [review]
Approving with a few minor suggestions

**Pull request merged**
